File: campus-navigate-app/expo-app/src/tiled/model/routePathModel.ts
import type { OverlayRouteNode } from '../type';
import { TEST_ROUTE_NODE_IDS } from './demoMapConfig';
import { createPathSegmentsFromPoints } from './pathSegmentModel';

export function createRoutePath(
  nodes: OverlayRouteNode[],
  nodeIds: readonly string[] = TEST_ROUTE_NODE_IDS,
) {
  const nodeById = new Map(nodes.map((node) => [node.nodeId, node]));
  const path: OverlayRouteNode[] = [];

  for (const nodeId of nodeIds) {
    const node = nodeById.get(nodeId);
    if (!node) {
      throw new Error(`Missing test route node: ${nodeId}`);
    }
    path.push(node);
  }

  return path;
}

export function createRoutePathSegments(
  nodes: OverlayRouteNode[],
  nodeIds: readonly string[] = TEST_ROUTE_NODE_IDS,
) {
  return createPathSegmentsFromPoints(createRoutePath(nodes, nodeIds));
}

File: campus-navigate-app/flutter_app/lib/domain/tiled/map/route_path_model.dart
import 'package:indoor_navigation/domain/config/app_config.dart';
import 'package:indoor_navigation/domain/tiled/map/path_segment_model.dart';
import 'package:indoor_navigation/domain/tiled/tiled_models.dart';

List<OverlayRouteNode> createRoutePath(
  List<OverlayRouteNode> nodes, {
  List<String> nodeIds = testRouteNodeIds,
}) {
  final nodeById = <String, OverlayRouteNode>{
    for (final node in nodes) node.nodeId: node,
  };
  final path = <OverlayRouteNode>[];

  for (final nodeId in nodeIds) {
    final node = nodeById[nodeId];
    if (node == null) {
      throw StateError('Missing test route node: $nodeId');
    }
    path.add(node);
  }

  return path;
}

List<OverlayPathSegment> createRoutePathSegments(
  List<OverlayRouteNode> nodes, {
  List<String> nodeIds = testRouteNodeIds,
}) {
  return createPathSegmentsFromPoints(createRoutePath(nodes, nodeIds: nodeIds));
}
